From 7c021d1a64ae49439c0a0011c4a941053a448f79 Mon Sep 17 00:00:00 2001
From: Du Yilong <duyilong@wxiat.com>
Date: Tue, 7 Jun 2022 17:42:12 +0800
Subject: [PATCH] sw64: add set time support for hypervisor based rtc
Patch-mainline: NO, OTHERS
References: openEuler-22.03-LTS
Modified-by-SEL: No


Sunway inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I56WV8

--------------------------------

Calculate the offset between target time and current time, then
add it to current time in read_time method. with this support,
writable requirements of rtc clock of guest os can be met.

Signed-off-by: Du Yilong <duyilong@wxiat.com>

Signed-off-by: Gu Zitao <guzitao@wxiat.com>
Acked-by: Xie XiuQi <xiexiuqi@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 drivers/rtc/rtc-sw64-virt.c | 24 +++++++++++++++++++++++-
 1 file changed, 23 insertions(+), 1 deletion(-)

diff --git a/drivers/rtc/rtc-sw64-virt.c b/drivers/rtc/rtc-sw64-virt.c
index 549d2e2d8a01..23c93d7ddbae 100644
--- a/drivers/rtc/rtc-sw64-virt.c
+++ b/drivers/rtc/rtc-sw64-virt.c
@@ -14,18 +14,40 @@
 #include <linux/platform_device.h>
 
 #define RTC_IO_ADDR (0x804910000000ULL)
+unsigned long vtime_old, vtime_new;
 
 static int sw64_virt_read_time(struct device *dev, struct rtc_time *tm)
 {
 	unsigned long *ioaddr;
+	unsigned long vtime_now;
+	long vtime_offset;
 
 	ioaddr = ioremap(RTC_IO_ADDR, sizeof(long));
-	rtc_time64_to_tm(*ioaddr, tm);
+	if (!vtime_new) {
+		rtc_time64_to_tm(*ioaddr, tm);
+	} else {
+		vtime_now = *ioaddr;
+		vtime_offset = vtime_new - vtime_old;
+		vtime_now += vtime_offset;
+		rtc_time64_to_tm(vtime_now, tm);
+	}
+	return 0;
+}
+
+static int sw64_virt_set_time(struct device *dev, struct rtc_time *tm)
+{
+	unsigned long *ioaddr;
+
+	ioaddr = ioremap(RTC_IO_ADDR, sizeof(long));
+	vtime_old = *ioaddr;
+
+	vtime_new = rtc_tm_to_time64(tm);
 	return 0;
 }
 
 static const struct rtc_class_ops rtc_sw64_virt_ops = {
 	.read_time	= sw64_virt_read_time,
+	.set_time	= sw64_virt_set_time,
 };
 
 static int __init rtc_sw64_virt_probe(struct platform_device *pdev)
-- 
2.34.1

